### ISSUE TITLE
Consolidate account-related shortcodes in a single shortcode

### ISSUE THREAD

**Issue by ericnicolaas** [Charitable/Charitable#144]
**The problem**

At the moment, we already have the following shortcodes related to accounts:

- `charitable_profile`
- `charitable_login`
- `charitable_registration`
- `charitable_submit_campaign` - Ambassadors
- `charitable_my_campaigns` - Ambassadors
- `charitable_creator_donations` - Ambassadors

Looking at some future functionality that will be added, the following could easily be added as shortcodes: 

- `charitable_forgot_password` - #89 
- `charitable_donation_history` - #14 
- Subscriptions shortcode in Recurring Donations - @helgatheviking 

Those are just the ones I can think of at the moment. It's not hard to imagine the list growing. The result is that getting Charitable setup, particularly when you have a few extensions installed, gets increasingly cumbersome, with lots of room for error. For every shortcode, you have to create a new page with the shortcode, and then (for most) you need to go to Charitable > Settings and pick the page you just created.

**The goal**

I would love to hear from others, but at this point I think the ideal solution is to have a single shortcode, with all of the other shortcodes replaced by endpoints. `[charitable_profile]` would be a good candidate. So the process for a new user would be to do the following: 

1. Create a new page with a `[charitable_profile]` shortcode.
2. Go to Charitable > Settings and nominate the page as the central account/profile/user management page.

Then, using WordPress' built-in ability to create endpoints, we organize the rest of the functionality based on that one page. You might end up with the following endpoints: 

- `domain.com/profile/` - Profile form. If user is not logged in, display login and (optionally) registration form.
- `domain.com/profile/donations/` - List of donations by the user.
- `domain.com/profile/campaigns/` - List of campaigns by the user (Ambassadors).
- `domain.com/profile/submit-campaign/` - Campaign submission form (Ambassadors).
- `domain.com/profile/donations-received/` - List of donations _received_ by the user (Ambassadors).
- `domain.com/profile/subscriptions/` - List of recurring donations for the user (Recurring Donations).

With this setup, the new endpoints are automatically available when you install an extension like Ambassadors or Recurring Donations. As long as you have that initial `[charitable_profile]` page set up (side note: it would be nice if users have the option of completely leaving out the account management side of things), you don't need to create any new pages.

That said, I think we can also provide a way for users to create a separate page for particular functionality, so that they can override the default endpoints system I have described above. For example, if you would like to have your campaign submission form listed at `domain.com/submit-campaign`, you would simply create a page with the `[charitable_submit_campaign]` shortcode.  

**Backwards compatibility & endpoint flexibility**

I am not proposing the removal of any of the existing shortcodes. Instead, I would suggest that the above `/profile/endpoint` system could be the _default_ setup, with the option for users to replace a particular endpoint with a page instead. For example, if you decide you want the URL for your campaign submission form to be `domain.com/submit-campaign`, you would create a page with the `[charitable_submit_campaign]` shortcode and then nominate that page as the campaign submission page. 

This is basically how we currently handle the Donation Receipt page. It works immediately out of the box with a preset endpoint, but you can create a page with a shortcode to replace that endpoint with your own preferred URL.

By retaining the existing shortcodes, we can avoid backwards compatibility issues. If you previously used Charitable with any of the shortcodes listed above, the page you used previously will continue to be used. 

**The road forward**

First of all, I want to hear feedback about my proposal above. 

I imagine that this might require a multi-stage process to get to the goal above. There are a couple of immediate questions that need answering if we move ahead with the above:

- How do we go about integration the "Forgot Password" functionality added by added by @rafecolton in #89? 
- How do we built out a donation history functionality (#14)?

Please comment and let me hear your thoughts.

**Comment by helgatheviking** [comment]
ENDPOINTS!! :+1:  

You can also let people re-name the endpoints with an admin setting.

![image](https://cloud.githubusercontent.com/assets/507025/16492171/7af7997c-3ea6-11e6-8554-46832c0019fb.png)

**Comment by ericnicolaas** [comment]
Yes, exactly. The WooCommerce approach is really my inspiration here.

**Comment by ericnicolaas** [comment]
Plus iThemes Exchange. They have a nice way of using automatic defaults that can be overridden through individual pages.

**Comment by helgatheviking** [comment]
I think that's a pretty powerful approach and prevents you from having tons of pages with nothing but shortcodes.

**Comment by ericnicolaas** [comment]
Here is how iThemes Exchange handles this: 

<img width="1018" alt="screenshot 2016-07-04 14 09 38" src="https://cloud.githubusercontent.com/assets/676582/16550943/f6dbcc64-41f0-11e6-9e8e-a25c5b5e305f.png">

The "Exchange" page type is basically the default, automatic option, but you can select "WordPress" from the dropdown and then nominate a page. It also then pops up with a note about the shortcode you should use on your page.

<img width="1012" alt="screenshot 2016-07-04 14 09 19" src="https://cloud.githubusercontent.com/assets/676582/16550944/f878ccd4-41f0-11e6-9797-899ade881e2a.png">

This is pretty close to what I envision for Charitable.

**Comment by rafecolton** [comment]
@ericnicolaas this approach looks very clear and straightforward 👍

**Comment by ericnicolaas** [comment]
In the interests of trimming down the size of 1.5.0, I'm going to punt this to 1.6.0 and instead focus on getting the groundwork done with #306 in 1.5.0.
